src/Microsoft.CodeQuality.Analyzers/UnitTests/QualityGuidelines/AvoidInfiniteRecursionTests.cs

```diff
@@ -0,0 +1,153 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Testing;
+using Xunit;
+using VerifyCS = Test.Utilities.CSharpCodeFixVerifier<
+    Microsoft.CodeQuality.Analyzers.QualityGuidelines.AvoidInfiniteRecursion,
+    Microsoft.CodeAnalysis.Testing.EmptyCodeFixProvider>;
+using VerifyVB = Test.Utilities.VisualBasicCodeFixVerifier<
+    Microsoft.CodeQuality.Analyzers.QualityGuidelines.AvoidInfiniteRecursion,
+    Microsoft.CodeAnalysis.Testing.EmptyCodeFixProvider>;
+
+namespace Microsoft.CodeQuality.Analyzers.UnitTests.QualityGuidelines
+{
+    public class AvoidInfiniteRecursionTests
+    {
+        [Fact]
+        public async Task PropertySetterRecursion_Diagnostic()
+        {
+            await VerifyCS.VerifyAnalyzerAsync(@"
+public class A
```

Comment:
```csharp
public class A
{
    private bool _flag;

    public int Abc
    {
        set
        {
            if (_flag)
            {
                 _flag = false;
                 this.Abc = value;
            }
        }
    }
}
```

This looks like possible false positive as user is guarding the infinite recursion explicitly with additional logic. Not flagging this case could lead to bunch of false negatives, so I am in two minds here. I am wondering if we can generate a separate and less assertive diagnostic message when the setter is parented by some conditional code OR always use a less assertive message for this diagnostic as we never know what crazy kind of code people write and guard infinite recursion. @sharwell thoughts?

Reply:
I like the idea of 2 "levels" of message. Although I am wondering if the conditional cases would be less important even if possibly less frequent.

Reply:
@mavasani In case we go for the 2 level of messages, I think it could be nice to start raising on lambda and local functions with this "uncertain" message, WDYT?